src/appleseed/foundation/math/knn/knn_anyquery.h

```diff
@@ -0,0 +1,397 @@
+    //   us to compute an initial maximum search distance which that speed up the real
+    //   search later on.
+    //
+
+    // Start at the root.
+    node = nodes;
+
+    FOUNDATION_KNN_QUERY_STATS(++fetched_node_count);
+
+    // In any case, we must stop when we reach a leaf node.
+    while (node->is_interior())
+    {
+        const size_t split_dim = node->get_split_dim();
+        const ValueType split_abs = node->get_split_abs();
+        const ValueType split_dist = query_point[split_dim] - split_abs;
+
+        // Figure out which child node to follow (the one that contains the query point).
+        // Points on the split plane belong to the right child node.
+        const size_t follow_index = split_dist >= ValueType(0.0) ? 1 : 0;
+        const NodeType* APPLESEED_RESTRICT follow_node = nodes + node->get_child_node_index() + follow_index;
+
+        FOUNDATION_KNN_QUERY_STATS(++fetched_node_count);
+
+        // Stop at the parent node if the child node contains too few points.
+        if (follow_node->get_point_count() < IdealLeafSize)
+            break;
+
+        // Continue with the child node.
+        node = follow_node;
+    }
+
+    //
+    // Step 2:
+    //
+    //   Check if one of the points in this node (if it's a leaf node) or below this
+    //   node (if it's an interior node) is within the query distance.
+    //
+
+    {
+        FOUNDATION_KNN_QUERY_STATS(++visited_leaf_count);
+
+        const VectorType* APPLESEED_RESTRICT point_ptr = points + node->get_point_index();
+        const VectorType* APPLESEED_RESTRICT point_end = point_ptr + node->get_point_count();
+
+        while (point_ptr < point_end)
+        {
+            FOUNDATION_KNN_QUERY_STATS(++tested_point_count);
+
+            const ValueType square_dist = square_distance(*point_ptr++, query_point);
+
+            if (square_dist <= query_max_square_distance)
+                goto found;
+        }
+    }
+
+    //
+    // Step 3:
+    //
+    //   Traverse again the tree, starting at the top, but this time we will push
+    //   nodes that we don't visit (but that potentially contain candidate points)
+    //   to a partially-ordered stack.
+    //
+
+#define ORDER_NODE_ENTRIES(LhsIndex, RhsIndex)                                                  \
+    if (node_queue[node_queue_size - (LhsIndex)].m_dvec_square_norm <                           \
+        node_queue[node_queue_size - (RhsIndex)].m_dvec_square_norm)                            \
+    {                                                                                           \
+        const NodeEntry tmp = node_queue[node_queue_size - (LhsIndex)];                         \
+        node_queue[node_queue_size - (LhsIndex)] = node_queue[node_queue_size - (RhsIndex)];    \
+        node_queue[node_queue_size - (RhsIndex)] = tmp;                                         \
+    }
+
+    const size_t NodeQueueSize = 128;
```

Comment:
`constexpr` can be faster (compilewise) here than `const`.

Reply:
In pratice it will be the same. Also `constexpr` doesn't guarantee that the value will be a pure compile-time one, just that it _can_ be used at compile-time.

In any case, you're right, the right semantic here is `constexpr` since we want to express a true constant and not a variable that is merely const. We need to progressively replace `const` by `constexpr` in the code base.